### ISSUE TITLE
Launching apps take more than 10 min

### ISSUE THREAD

**Issue by atifwaqar** [Starcounter/Starcounter#3494]
I am using 3 Starcounter apps and 4th app is our app (AdFenix). and the apps are launched in following sequence:
1. Launcher
2. UserAdmin
3. Signin
4. AdFenix

When i run my batch file to run the apps, Launcher app takes more than 10 min (sometimes even more than 20 min). Any idea what can be the issue? and is it possible to speed up the launch process?

My DB size (personal folder size) is 27 GB after cleaning the logs.

**Comment by per-samuelsson** [comment]
@atifsheikh 

We are actually in the middle of a sprint where we are reviewing the application launch cycle, both for stability issues and performance. So this feedback suit us well.

With that being said, wow, those timings seem ... well, really out of line. I'm quite surprised you experience something *that* bad.

Let me get back to you on this later today, possibly with some instructions on how to collect some diagnostics.

**Comment by dan31** [comment]
Hi @atifsheikh! Please, as usual, refer to an exact version your are running. Thanks!

**Comment by dan31** [comment]
@atifsheikh, did it happen after migrating to recent version or this was there around in 2.1.177 as well?

**Comment by atifsheikh** [comment]
@dan31 this is not version specific, we had this issue in 2.1.177 and even in the SC Stable Version.

**Comment by per-samuelsson** [comment]
@atifsheikh 

In #3481, you can see some of the efforts I started some days ago.

I'll take a spin out of those first, and will ask you to try this thing for me on your environment.

1. Create a new folder
2. In that folder, create a file **timeboot.bat** and paste in the below content.
3. Run it with `timeboot [database]`
4. Copy content of **timeboot.txt** and paste that here.

Example of running in default database:

> C:\Users\You\NewDir\timeboot default

**timeboot.bat**
```bat
@echo off

set db=%1

IF NOT EXIST app.cs (
  staradmin +
)

set SC_TIMESTAMPS_FILE_PATH=
IF EXIST timeboot.txt (
  del timeboot.txt
)

staradmin kill all

set SC_TIMESTAMPS_FILE_PATH=%cd%\timeboot.txt
staradmin start server

:: We are not interested in results from booting up
:: admin server. Delete what was emitted up until now.
del timeboot.txt

set start=%TIME%
star -d=%db% --logsteps app.cs
set stop=%TIME%
set SC_TIMESTAMPS_FILE_PATH=

ECHO Frame: %start% - %stop%
ECHO Diagnostics also in timeboot.txt
```

**Comment by alemoi** [comment]
:+1: to @per-samuelsson for the good measurement way.

**Comment by atifwaqar** [comment]
@per-samuelsson 
I ran the timeboot.bat file in my localmachine and as soon i launched it, it closed all my running apps.
so i cant run this batch file on any live server in daytime, i will test this in evening, and will update you afterwards. 
Thanks

**Comment by per-samuelsson** [comment]
> I ran the timeboot.bat file in my localmachine and as soon i launched it, it closed all my running apps.

Sorry, I should have been clear about that. It's caused by `staradmin kill all`.

> i will test this in evening, and will update you afterwards. 

Thanks a lot. We really appreciate that.

**Comment by atifwaqar** [comment]
Sorry for the delay, Luckily our code was not crashing and we did not need to restart the apps. 

```
Starcounter.Server.Commands: Executing 'Starting database default.' (fd1696d5-ef04-448a-b0b2-2eb2512528bb): 05.34.02.213
Starcounter.Server.Commands: Begin task 'Starting database process': 05.34.02.218
Starcounter.Server.Commands: End task (cancelled=False): 05.34.02.251
Starcounter.Server.Commands: Begin task 'Starting log writer process': 05.34.02.251
Starcounter.Server.Commands: End task (cancelled=False): 05.34.02.258
Starcounter.Server.Commands: Begin task 'Starting code host process': 05.34.02.259
Starcounter.Server.Commands: End task (cancelled=False): 05.34.02.277
Starcounter.Server.Commands: Begin task 'Assuring code host is ready for service': 05.34.02.278
Starcounter.Host: Started sccode Main(): 05.34.02.298
Starcounter.Host: Bootstrap Main() started.: 05.34.02.317
Starcounter.Host: Exception factory installed.: 05.34.02.337
Starcounter.Host: Command line parsed.: 05.34.02.383
Starcounter.Host: Configuration loaded.: 05.34.02.385
Starcounter.Host: Assured no other process with the same name.: 05.34.02.388
Starcounter.Host: Global memory allocated.: 05.34.02.388
Starcounter.Host: Kernel memory configured.: 05.34.02.392
Starcounter.Host: Logging configured.: 05.34.02.412
Starcounter.Host: BMX manager initialized.: 05.34.02.418
Starcounter.Host: Host configured.: 05.34.02.420
Starcounter.Host: Scheduler configured.: 05.34.02.443
Starcounter.Host: Apps bootstraper initialized.: 05.34.03.802
Starcounter.Host: Database configured.: 05.34.03.810
Starcounter.Host: Database connected.: 05.36.30.791
Starcounter.Host: Query module initiated.: 05.36.30.844
Starcounter.Host: Setup completed.: 05.36.30.845
Starcounter.Host: Scheduler started.: 05.36.30.848
Starcounter.Host: Network gateway connected.: 05.36.30.848
Starcounter.Host: App domain configured.: 05.36.30.848
Starcounter.Host: Server command handlers registered.: 05.36.32.611
Starcounter.Host.Loader: Package created.: 05.36.32.881
Starcounter.Host.Loader: Package started.: 05.36.32.891
Starcounter.Host.Loader: Runtime meta-data tables were created and populated with initial data.: 05.36.33.004
Starcounter.Host.Loader: CLR view meta-data were deleted on host start.: 05.36.33.031
Starcounter.Host.Loader: Properties and columns were populated for the given meta-types.: 05.36.33.053
Starcounter.Host.Loader: Types of database schema checked and updated.: 05.36.33.074
Starcounter.Host.Loader: Type definitions registered.: 05.36.33.079
Starcounter.Host.Loader: Columns of database schema checked and updated.: 05.36.33.079
Starcounter.Host.Loader: Query module schema information updated.: 05.36.33.112
Starcounter.Host.Loader: System type specifications initialized.: 05.36.33.239
Starcounter.Host.Loader: CLR view meta-data were populated for the given types.: 05.36.34.027
Starcounter.Host.Loader: Internal handlers were registered.: 05.36.35.132
Starcounter.Host.Loader: Processing completed.: 05.36.35.132
Starcounter.Host.Loader: Package processed.: 05.36.35.133
Starcounter.Host: Base package loaded.: 05.36.35.133
Starcounter.Host: Start completed.: 05.36.35.134
Starcounter.Host: Booting completed.: 05.36.35.137
Starcounter.Server.Commands: End task (cancelled=False): 05.36.37.992
Starcounter.Server.Commands: Executing completed (fd1696d5-ef04-448a-b0b2-2eb2512528bb): 05.36.38.047
Starcounter.Server.Commands: Executing 'Starting app in default' (05d0d099-f955-4621-9212-d57a21be67a5): 05.36.38.797
Starcounter.Server.Commands: Begin task 'Preparing user executables and files': 05.36.38.807
Starcounter.Server.Commands: Weaving completed.: 05.36.51.957
Starcounter.Server.Commands: End task (cancelled=False): 05.36.51.963
Starcounter.Server.Commands: Begin task 'Loading executable in code host': 05.36.51.963
Starcounter.Host.Loader: Loader started.: 05.36.52.192
Starcounter.Host.Loader: Input verified and assembly resolver updated.: 05.36.52.318
Starcounter.Host.Loader: Schema verified and loaded.: 05.36.52.614
Starcounter.Host.Loader: Target assembly loaded.: 05.36.52.717
Starcounter.Host.Loader: Package created.: 05.36.52.718
Starcounter.Host.Loader: Package started.: 05.36.52.779
Starcounter.Host.Loader: Types of database schema checked and updated.: 05.36.52.853
Starcounter.Host.Loader: Type definitions registered.: 05.36.52.854
Starcounter.Host.Loader: Columns of database schema checked and updated.: 05.36.52.854
Starcounter.Host.Loader: Query module schema information updated.: 05.36.53.326
Starcounter.Host.Loader: CLR view meta-data were populated for the given types.: 05.36.53.462
Starcounter.Host.Loader: Entry point executed.: 05.36.53.739
Starcounter.Host.Loader: Processing completed.: 05.36.53.740
Starcounter.Host.Loader: Package processed.: 05.36.53.740
Starcounter.Server.Commands: Code host Exec request processed.: 05.36.53.743
Starcounter.Server.Commands: Database app registered.: 05.36.53.743
Starcounter.Server.Commands: Database status updated.: 05.36.54.804
Starcounter.Server.Commands: End task (cancelled=False): 05.36.54.804
Starcounter.Server.Commands: Executing completed (05d0d099-f955-4621-9212-d57a21be67a5): 05.36.54.804

```

**Comment by alemoi** [comment]
Thank you @atifwaqar! Wow, for me it seems that every single step is extremely slow! And some steps are super extremely slow! Reminds me of situation that @mtornwall had, but with gateway. May it be related that scservice and subsequent processes start with some super low priority? What the overall machine performance? Usage of CPU when you try to start Starcounter? Feels like its 100% used during that time.

**Comment by per-samuelsson** [comment]
Thanks a lot for helping us with this @atifwaqar!

@alemoi 

> Wow, for me it seems that every single step is extremely slow! And some steps are super extremely slow!

Well, they are clearly slower than those from our machines, but the *relative* slowness is very compatible to what we found. Booting up an empty code host is like 90% of the total time, and out of that, this step

> Starcounter.Host: Database configured.: 05.34.03.810
Starcounter.Host: Database connected.: 05.36.30.791

is like 90%. That's 2,5 minutes the code host just wait for database to become "ready".

And in comparison, weaving is ~13 seconds

> Starcounter.Server.Commands: Begin task 'Preparing user executables and files': 05.36.38.807
Starcounter.Server.Commands: Weaving completed.: 05.36.51.957

on that same environment, and that relationship is really really off (as we've aimed historically to have weaver be the bottleneck).

**Comment by mtornwall** [comment]
My two cents:

I've seen similar startup times (perhaps not as bad as ten minutes, though) while conducting scalability tests.

My experience has been that starting a large database doesn't take take that long *in itself*. For example, the database on my development machine has tens of millions of records, and it generally starts in less than ten seconds. Instead I've tended to experience these slow startups after concluding a very long write-intensive test.

I've always assumed (and perhaps I'm out riding my bike, to [ab]use a transliterated Swedish idiom :-)) that the database is sifting through the logs and perhaps generating a checkpoint, and that this is the cause of the apparent slow startup. But I'm very interested in hearing an official explanation. :-)

**Comment by alemoi** [comment]
> but the relative slowness is very compatible to what we found. 

Yes, that's correct. Everything becomes super slow. The same I remember from the gateway startup log that @mtornwall provided.

**Comment by atifwaqar** [comment]
If you guys need that (slow) server's DB dump to check if it takes the same time on your machine, I can help you with that.

**Comment by per-samuelsson** [comment]
> If you guys need that (slow) server's DB dump to check if it takes the same time on your machine, I can help you with that.

I downloaded it yesterday, and the apps as well. And I've set up my environment to run it. Right now, I experience some error I'm not sure how to get past though (#3607), so it's a little bit on hold. Just wanted to give you some feedback on the status.

BTW @atifwaqar, do you know if you made sure no Starcounter processes where running when you did copy that `Starcounter\Personal` file structure? I think it would have failed if you *had not* done so, but I'm just asking any way.

**Comment by atifwaqar** [comment]
> BTW @atifwaqar, do you know if you made sure no Starcounter processes where running when you did copy that Starcounter\Personal file structure? I think it would have failed if you had not done so, but I'm just asking any way.

If you mean my apps were running while I copied the DB, then yes all of my apps were running and i made a copy of my personal folders, then i zipped the copied personal folders. 
I always get error on 2 files when I copy personal folder with all apps running. and both files are in the following folder : 
`C:\Users\Administrator\Documents\Starcounter\Personal\Data\Default\Default-20160428T183646289\`
and both are log files, the file names look like `default.0000000013.log`

The reason I could not stop my apps before starting the copy process is that the server is a live server, and we can not afford to have the apps down.

But do let me know if you need the DB backup with all apps closed. I'll try to make the backup in the minimum traffic hours.

**Comment by per-samuelsson** [comment]
> If you mean my apps were running while I copied the DB, then yes all of my apps were running and i made a copy of my personal folders, then i zipped the copied personal folders. 

Then yes, the database was running when you made the copy. And that's probably why I can't run it on my end, and possibly the reason for #3607.

> But do let me know if you need the DB backup with all apps closed. I'll try to make the backup in the minimum traffic hours.

I think we do unfortunately. I'm not aware of any other way to safely share a database. But I do understand the problematic nature of that, when we are talking about live environments.

Since this is not a showstopper, I'd suggest you wait a little to do so. Meanwhile, I'll ping @eriohl to just confirm that I'm right in my guess that there is no other way than stopping the database. Also @eriohl, my guess about that this is the cause for #3607 - do you think it's likely enough to be so to close that issue?

**Comment by eriohl** [comment]
> Since this is not a showstopper, I'd suggest you wait a little to do so. Meanwhile, I'll ping @eriohl to just confirm that I'm right in my guess that there is no other way than stopping the database.

Yes. If you want the complete database (to this I would like to add that this has been addressed in a later version) or what to make sure that you get a working database.

Although you could get a working database anyway if you are lucky and there is no conflict with a checkpoint. So you could simply try copying it again and if the new copy works.

> Also @eriohl, my guess about that this is the cause for #3607 - do you think it's likely enough to be so to close that issue?

Very likely yes. A conflict with a checkpoint explains the two volatile images.

**Comment by per-samuelsson** [comment]
Thanks for the feedback @eriohl.

> Although you could get a working database anyway if you are lucky and there is no conflict with a checkpoint. So you could simply try copying it again and if the new copy works.

Then it's up to you @atifwaqar. For me, I can test again and again with new databases until we find one that works, if you don't want to stop the server to do the copy. I'll just act on your instructions on this. And in your timeframe (meanwhile, I'm working something else).

> Very likely yes. A conflict with a checkpoint explains the two volatile images.

OK. Then I'll close that one with that reason as the probable cause.

**Comment by eriohl** [comment]
> Although you could get a working database anyway if you are lucky and there is no conflict with a checkpoint. So you could simply try copying it again and if the new copy works.

Actually. Forget that comment. I thought about it a moment longer and I must conclude that although possible it is rather unlikely that the copy will work.

**Comment by eriohl** [comment]
But is the actual database image actually required here. If you just need the database data another option is of course to do a SQL dump (unload).

**Comment by atifwaqar** [comment]
@per-samuelsson I will make a copy of DB on any night of coming weekend then.

**Comment by per-samuelsson** [comment]
> But is the actual database image actually required here. If you just need the database data another option is of course to do a SQL dump (unload).

We are not sure of that, and actually, it's most likely that it's something with the environment, so the kind of representation is possibly not even important. But we wanted to try just mirroring the exact data setup from their server to one of ours, to see if numbers where similar.

**Comment by per-samuelsson** [comment]
@atifwaqar 

I've got that new database and set up an environment with it now, and I don't get the inconsistency error as I did with the last database. :+1: 

However, I seem not to be able to run it on my machine because of #3630. We have to consider this some internally, and will get back to you on that.

**Comment by dan31** [comment]
We have discussed that it is not recommended to kill all starcounter processes just to restart apps: this will require database to be started, which, in case of tens of gb-s in Db, require quite some time. Only stop apps, not kill Starcounter.
